sdks/python/apache_beam/io/mongodbio.py

```diff
@@ -139,50 +143,77 @@ def __init__(self,
     self.filter = filter
     self.projection = projection
     self.spec = extra_client_params
-    self.doc_count = self._get_document_count()
-    self.avg_doc_size = self._get_avg_document_size()
-    self.client = None
 
   def estimate_size(self):
-    return self.avg_doc_size * self.doc_count
+    with MongoClient(self.uri, **self.spec) as client:
+      size = client[self.db].command('collstats', self.coll).get('size')
+      if size is None or size <= 0:
+        raise ValueError('Collection %s not found or total doc size is '
+                         'incorrect' % self.coll)
+      return size
 
   def split(self, desired_bundle_size, start_position=None, stop_position=None):
     # use document cursor index as the start and stop positions
     if start_position is None:
-      start_position = 0
+      epoch = datetime.datetime(1970, 1, 1)
+      start_position = objectid.ObjectId.from_datetime(epoch)
     if stop_position is None:
-      stop_position = self.doc_count
+      last_doc_id = self._get_last_document_id()
+      # add one sec to make sure the last document is not excluded
+      last_timestamp_plus_one_sec = (last_doc_id.generation_time +
+                                     datetime.timedelta(seconds=1))
+      stop_position = objectid.ObjectId.from_datetime(
+          last_timestamp_plus_one_sec)
 
-    # get an estimate on how many documents should be included in a split batch
-    desired_bundle_count = desired_bundle_size // self.avg_doc_size
+    desired_bundle_size_in_mb = desired_bundle_size // 1024 // 1024
+    split_keys = self._get_split_keys(desired_bundle_size_in_mb, start_position,
+                                      stop_position)
 
     bundle_start = start_position
-    while bundle_start < stop_position:
-      bundle_end = min(stop_position, bundle_start + desired_bundle_count)
-      yield iobase.SourceBundle(weight=bundle_end - bundle_start,
+    for split_key_id in split_keys:
+      bundle_end = min(stop_position, split_key_id)
+      if bundle_start is None and bundle_start < stop_position:
+        return
+      yield iobase.SourceBundle(weight=desired_bundle_size_in_mb,
                                 source=self,
                                 start_position=bundle_start,
                                 stop_position=bundle_end)
       bundle_start = bundle_end
+    # add range of last split_key to stop_position
+    if bundle_start < stop_position:
+      yield iobase.SourceBundle(weight=desired_bundle_size_in_mb,
+                                source=self,
+                                start_position=bundle_start,
+                                stop_position=stop_position)
 
   def get_range_tracker(self, start_position, stop_position):
     if start_position is None:
-      start_position = 0
+      epoch = datetime.datetime(1970, 1, 1)
```

Comment:
When can start_position and stop_position be None in get_range_tracker?

Reply:
My understanding is when the first split happens.